tests/test_cache.py

```diff
@@ -0,0 +1,105 @@
+from cachecontrol.caches.redis_cache import RedisCache
+from canonicalwebteam.http import CachedSession
+from datetime import timedelta
+
+file_cache_directory = ".testcache"
+
+
+def create_redis_connection_pool(
+    redis_port="6379", redis_host="localhost", redis_db=0
+):
+    return redis.ConnectionPool(host=redis_host, port=redis_port, db=redis_db)
+
+
+class TestCachedSession(unittest.TestCase):
+    def tearDown(self):
+        try:
+            shutil.rmtree(file_cache_directory)
+        except:
+            pass
+
+    @httpretty.activate
+    def test_custom_heuristic(self):
+        httpretty.register_uri(
+            httpretty.GET,
+            "https://httpbin.org/ip",
+            body='{"origin": "127.0.0.1"}',
+        )
+        session = CachedSession(
+            fallback_cache_duration=10,
+            file_cache_directory=file_cache_directory,
+        )
+        resp = session.get("https://httpbin.org/ip")
+
+        for h in resp.headers:
+            if h == "cache-control":
+                print(resp.headers[h])
+
+        self.assertIn(
+            "Automatically cached for 0:00:10", resp.headers["warning"]
+        )
+
+    @httpretty.activate
+    def test_default_heuristic(self):
+        httpretty.register_uri(
+            httpretty.GET,
+            "https://httpbin.org/ip",
+            body='{"origin": "127.0.0.1"}',
+        )
+        session = CachedSession(file_cache_directory=file_cache_directory)
+        resp = session.get("https://httpbin.org/ip")
+
+        self.assertIn(
+            "Automatically cached for 0:00:05", resp.headers["warning"]
+        )
+
+    @httpretty.activate
+    def test_cache_control_max_age_overwrites_custom_heuristic(self):
+        httpretty.register_uri(
+            httpretty.GET,
+            "https://httpbin.org/ip",
+            body='{"origin": "127.0.0.1"}',
+            adding_headers={"Cache-Control": "max-age=30"},
+        )
+        session = CachedSession(file_cache_directory=file_cache_directory)
+        resp = session.get("https://httpbin.org/ip")
+
+        self.assertNotIn("warning", resp.headers)
+        self.assertIn("max-age=30", resp.headers["cache-control"])
+
+    @httpretty.activate
+    def test_cache_control_no_cache_overwrites_custom_heuristic(self):
+        httpretty.register_uri(
+            httpretty.GET,
+            "https://httpbin.org/ip",
+            body='{"origin": "127.0.0.1"}',
+            adding_headers={"Cache-Control": "no-cache"},
+        )
+        session = CachedSession(file_cache_directory=file_cache_directory)
+        resp = session.get("https://httpbin.org/ip")
+
+        self.assertNotIn("warning", resp.headers)
+        self.assertIn("no-cache", resp.headers["cache-control"])
+
+    def test_file_cache(self):
+        session = CachedSession()
+        adapter = session.get_adapter("http://")
+        cache = adapter.cache
+        self.assertIsInstance(cache, FileCache)
```

Comment:
I would remove this test entirely - it's purely testing internals. It doesn't matter whether the "http://" adapter is a cache instance, as long as the module successfully caches responses.

Reply:
I have updated this test. The problem that I see here is that I again have knowledge in my Test about how the library encodes the data. 
[`CacheControl` is using msgpack](https://github.com/ionrock/cachecontrol/blob/master/cachecontrol/serialize.py#L72). These are again internals. 

I have tried to get a bit of decoupling by using `str()` on both the epoch (comparing the whole response would need proper deserialization) and the cached file.

Reply:
Also, I am not sure how to manipulate the file smartly without the knowledge of the library

Reply:
Oh yeah, this should work nicely. Thanks! I wont be able to manipulate data, and see that I am getting my modified response, but I think all necessary functionality will still be covered.